Add 'sendall' RPC née sweep

murchandamus opened the pull request:
Add sendall RPC née sweep

_Motivation_
Currently, the wallet uses a fSubtractFeeAmount (SFFO) flag on the
recipients objects for all forms of sending calls. According to the
commit discussion, this flag was chiefly introduced to permit sweeping
without manually calculating the fees of transactions. However, the flag
leads to unintuitive behavior and makes it more complicated to test
many wallet RPCs exhaustively. We proposed to introduce a dedicated
`sendall` RPC with the intention to cover this functionality.

Since the proposal, it was discovered in further discussion that our
proposed `sendall` rpc and SFFO have subtly different scopes of
operation.
• sendall:
  Use _given UTXOs_ to pay a destination the remainder after fees.
• SFFO:
  Use a _given budget_ to pay an address the remainder after fees.

While `sendall` will simplify cases of spending a given set of
UTXOs such as paying the value from one or more specific UTXOs, emptying
a wallet, or burning dust, we realized that there are some cases in
which SFFO is used to pay other parties from a limited budget,
which can often lead to the creation of change outputs. This cannot be
easily replicated using `sendall` as it would require manual
computation of the appropriate change amount.

As such, sendall cannot replace all uses of SFFO, but it still has a
different use case and will aid in simplifying some wallet calls and
numerous wallet tests.

_Sendall call details_
The proposed sendall call builds a transaction from a specific
subset of the wallet's UTXO pool (by default all of them) and assigns
the funds to one or more receivers. Receivers can either be specified
with a given amount or receive an equal share of the remaining
unassigned funds. At least one recipient must be provided without
assigned amount to collect the remainder. The `sendall` call will
never create change. The call has a `send_max` option that changes the
default behavior of spending all UTXOs ("no UTXO left behind"), to
maximizing the output amount of the transaction by skipping uneconomic
UTXOs. The `send_max` option is incompatible with providing a specific
set of inputs.

---
Edit: Replaced OP with latest commit message to reflect my updated motivation of the proposal.

murchandamus committed:
Add sendall RPC née sweep

_Motivation_
Currently, the wallet uses a fSubtractFeeAmount (SFFO) flag on the
recipients objects for all forms of sending calls. According to the
commit discussion, this flag was chiefly introduced to permit sweeping
without manually calculating the fees of transactions. However, the flag
leads to unintuitive behavior and makes it more complicated to test
many wallet RPCs exhaustively. We proposed to introduce a dedicated
`sendall` RPC with the intention to cover this functionality.

Since the proposal, it was discovered in further discussion that our
proposed `sendall` rpc and SFFO have subtly different scopes of
operation.
• sendall:
  Use _specific UTXOs_ to pay a destination the remainder after fees.
• SFFO:
  Use a _specific budget_ to pay an address the remainder after fees.

While `sendall` will simplify cases of spending from specific UTXOs,
emptying a wallet, or burning dust, we realized that there are some
cases in which SFFO is used to pay other parties from a limited budget,
which can often lead to the creation of change outputs. This cannot be
easily replicated using `sendall` as it would require manual computation
of the appropriate change amount.

As such, sendall cannot replace all uses of SFFO, but it still has a
different use case and will aid in simplifying some wallet calls and
numerous wallet tests.

_Sendall call details_
The proposed sendall call builds a transaction from a specific subset of
the wallet's UTXO pool (by default all of them) and assigns the funds to
one or more receivers. Receivers can either be specified with a specific
amount or receive an equal share of the remaining unassigned funds. At
least one recipient must be provided without assigned amount to collect
the remainder. The `sendall` call will never create change. The call has
a `send_max` option that changes the default behavior of spending all
UTXOs ("no UTXO left behind"), to maximizing the output amount of the
transaction by skipping uneconomic UTXOs. The `send_max` option is
incompatible with providing a specific set of inputs.

JeremyRubin commented:
Interesting idea. Can this be exposed as a testing only RPC? Or do real users need it?

There are privacy implications of using sweep as well -- if it's just to be used for e.g. taking an old potentially compromised wallet and migrating to a new one is it better to do a sweep that splits into a number of normal-ish looking txns?

jamesob commented:
Cool, concept ACK

Tests look nice

Xekyo commented:
Subtract fee from amount has historically been used for sweeping and there exist some issues on this repo that illustrate that there is usage.

You could achieve a "normal" looking transaction by specifying two recipients and setting an amount on one of them. There are examples to illustrate this usage in the RPC help text.

brunoerg commented:
Concept ACK

glozow commented:
Concept ACK, I prefer this to #23534.

w0xlt commented:
Concept ACK

shaavan commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1159,6 +1160,267 @@ RPCHelpMan send()
+            {"conf_target", RPCArg::Type::NUM, RPCArg::DefaultHint{"wallet -txconfirmtarget"}, "Confirmation target in blocks"},
+            {"estimate_mode", RPCArg::Type::STR, RPCArg::Default{"unset"}, std::string() + "The fee estimate mode, must be one of (case insensitive):\n"
+                        "       \"" + FeeModes("\"\n\"") + "\""},
+            {"fee_rate", RPCArg::Type::AMOUNT, RPCArg::DefaultHint{"not set, fall back to wallet fee estimation"}, "Specify a fee rate in " + CURRENCY_ATOM + "/vB."},
+            {"options", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED_NAMED_ARG, "",
+                {
+                    {"add_to_wallet", RPCArg::Type::BOOL, RPCArg::Default{true}, "When false, returns a serialized transaction which will not be added to the wallet or broadcast"},
+                    {"fee_rate", RPCArg::Type::AMOUNT, RPCArg::DefaultHint{"not set, fall back to wallet fee estimation"}, "Specify a fee rate in " + CURRENCY_ATOM + "/vB."},
```
The fee_rate is defined twice:
1. As an argument
2. As an option object

I presume this is an intentional design choice. But I can't seem to grasp the reason behind doing so.

shaavan replied:
> This is consistent with how all other transaction creating RPCs permit setting the `fee_rate`.

I checked with other RPCs where `"fee_rate", RPCArg::Type::AMOUNT,...` is used. It seems like the situation you are describing is true only for the `send` RPC (see https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpc/spend.cpp#L1050-L1062) and isn't a norm per se.

Other RPCs that uses fee_rate but doesn't define it twice are:
- sendtoaddress
- sendmany
- fundrawtransaction
- bumpfee_helper, and,
- walletcreatefundedpsbt

Xekyo replied:
Okay, I'll take another look at that. Thank you.

Xekyo replied:
After talking to a few people, my sentiment is that this is the expected behavior. In the minimal use case, the most important arguments are available as positional, making it easy to use. When people start using options, they may want to define everything as options, though, for consistency and readability.

shaavan commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1159,6 +1160,267 @@ RPCHelpMan send()
+            if (!pwallet) return NullUniValue;
+
+            UniValue options{request.params[4].isNull() ? UniValue::VOBJ : request.params[4]};
+            if (options.exists("conf_target") || options.exists("estimate_mode")) {
+                if (!request.params[1].isNull() || !request.params[2].isNull()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Pass conf_target and estimate_mode either as arguments or in the options object, but not both");
```
Two Questions:

1. If I pass the `conf_target` argument and `estimate_mode` under the options object, this will cause an error. But should there be an error in such a scenario?

2. `conf_target` and `estimate_mode` can be options objects, but they are not explicitly defined in the RPC help message under the options column as is done with the fee_rate object. Is it a deliberate decision?

Xekyo replied:
Good catch, thanks. I'll make sure that the `conf_target` and `estimate_mode` appear in the documentation of the options.

Regarding the error when `conf_target` and `estimate_mode` are passed in different ways, this is consistent with what `send` does.

shaavan replied:
> Regarding the error when conf_target and estimate_mode are passed in different ways, this is consistent with what send does.

I checked the code, and it seems like you are right. I can also see that the `sweep` RPC is modeled after the `send` RPC. However, I am not sure if it is an optimal behavior. I can't understand why the user is compelled to input both of these options as either options or as arguments. Why not allow them to pass the first `conf_target` in one way and `estimate_mode` in another. Similar to what is done with `fee_rate`.

shaavan commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1159,6 +1160,267 @@ RPCHelpMan send()
+                    {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "A bitcoin address which receives an equal share of the unspecified amount."},
+                    {"", RPCArg::Type::OBJ_USER_KEYS, RPCArg::Optional::OMITTED, "",
+                        {
+                            {"address", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "A key-value pair. The key (string) is the bitcoin address, the value (float or string) is the amount in " + CURRENCY_UNIT + ""},
```
nit:
I think it's better to replace `key-value` -> `key:value`, because that's how the pair is originally expressed in arguments.

Xekyo replied:
"key-value" is the prevalent way of describing this sort of datum.

shaavan replied:
> "key-value" is the prevalent way of describing this sort of datum.

I had checked with various RPC commands, and I think it is correct.

shaavan commented on test/functional/wallet_sweep.py:
```diff
@@ -0,0 +1,172 @@
+
+        self.clean_up()
+
+    def sweep_negative_effective_value(self):
```
I think you can add a `self.log.info(..)` statement here.

Xekyo replied:
Done

shaavan commented on test/functional/wallet_sweep.py:
```diff
@@ -0,0 +1,172 @@
+
+        dust_wallet.unloadwallet()
+
+    def sweep_with_sendmax(self):
```
Here also, a `self.log.info(..)` statement can be added.

Xekyo replied:
Done

shaavan commented on test/functional/wallet_sweep.py:
```diff
@@ -0,0 +1,172 @@
+        self.generate(self.nodes[0], 1)
+        assert_greater_than(self.wallet.getbalances()["mine"]["trusted"], 0)
+
+        sweep_tx_receipt = self.wallet.sweep(receivers=[self.return_addr_remainder], fee_rate=300, options={"sendmax": True})
```
Though it's not very difficult grasping the meaning of each section, I think it would be better if you added a comment:
```
sweep with sendmax enabled.
```
Before line 127. And statement:
```
sweep with sendmax disabled.
```
Before line 138.

Xekyo replied:
I've added comments as suggested.

shaavan reviewed (commented):
Concept ACK

The added `sweep()` RPC function looks good. I went through the code and couldn’t find any breaking points. The same goes for the added test, which is logically sound. Kudos for this amazing work, @Xekyo! :clinking_glasses: 

Here’s a screenshot of the displayed message when passing the command:
```
./src/bitcoin-cli -signet sweep
```
**Screenshot:**

![PR](https://user-images.githubusercontent.com/85434418/150640566-d2744855-89e5-4760-a7a8-bffe52d9ffcc.png)

The added test passed successfully, which intends that the added function works correctly. I shall do a manual test by creating a new wallet and sweeping it in some time. Meanwhile, I have some doubts and suggestions that might help make this PR even better.

brunoerg commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1159,6 +1160,267 @@ RPCHelpMan send()
+            }
+
+            if (complete) {
+                std::string err_string;
```
Is there a place where this string is used? I could see only this declaration.

Xekyo replied:
Thanks. I also removed it from send where it also wasn't used.

achow101 commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1159,6 +1160,267 @@ RPCHelpMan send()
+            }
+
+            bool rbf = pwallet->m_signal_rbf;
+            if (options.exists("replaceable")) {
```
`replaceable` is not documented

Xekyo replied:
Fixed, thanks.

achow101 commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1159,6 +1160,267 @@ RPCHelpMan send()
+                if (sendmax && fee_rate.GetFee(output.nInputBytes) > output.tx->tx->vout[output.i].nValue) {
+                    continue;
+                }
+                CTxIn input(output.tx->GetHash(), output.i);
```
Needs to handle `rbf` and set the right sequence number

```suggestion
                CTxIn input(output.tx->GetHash(), output.i, CScript(), rbf ? MAX_BIP125_RBF_SEQUENCE : CTxIn::SEQUENCE_FINAL);
```

Xekyo replied:
Added, thanks.

achow101 commented:
The GUI will need access to this functionality too. It would be nice if the sweeping logic was refactored into a function in `src/wallet/spend.cpp` so that the GUI can access it.

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1159,6 +1160,267 @@ RPCHelpMan send()
+RPCHelpMan sweep()
+{
+    return RPCHelpMan{"sweep",
+        "\nEXPERIMENTAL warning: this call may be changed in future releases.\n"
```
Removed the newline. (I was copying the style of `send` has one.)

Sjors commented:
I have no objection to a dedicated `sweep` RPC call, but I don't agree with the premise that subtracting fee from output is only useful for sweeping a full wallet. See https://github.com/bitcoin/bitcoin/pull/24142#issuecomment-1024145620

ryanofsky commented:
> I have no objection to a dedicated `sweep` RPC call, but I don't agree with the premise that subtracting fee from output is only useful for sweeping a full wallet. See [#24142 (comment)](https://github.com/bitcoin/bitcoin/pull/24142#issuecomment-1024145620)

Either approach seems fine to me, but I do think it is potentially more confusing to have two completely different RPCs for sending funds: "send" vs" sweep" than to have one sending option that means "Send approximate amount. I don't care about exact amount of BTC received, and am happy if it's a little less or a little more to economize on fees and avoid change." Obviously you shouldn't use this option if you are trying to send an exact amount to someone, but it seems like it would be generally useful whenever you are sending money to one of your own wallets, or exchange accounts, or paying for any service that can be incrementally topped up.

I know one of achow101's recent PR was adding more complexity to subtract from output implementation to try to do something to help sweeping, but I don't think the original (current?) semantics inherently had to add much complexity. It is true the code has been complex at different points but I think that was mostly a result of code shittiness and duplication, which have generally improved recently.

Sjors commented:
Indeed I suspect that a `sweep` RPC will be useless if we decide to keep subtracting fee from output functionality. In that case adding a sweep feature to the `send` RPC would make more sense.

w0xlt commented:
Concept ACK

I have no objections to this RPC either, but something like `bitcoin-cli -named sendtoaddress address="...." sweep=true` might be simpler for users instead of two commands to send funds.

achow101 commented:
> I have no objection to a dedicated `sweep` RPC call, but I don't agree with the premise that subtracting fee from output is only useful for sweeping a full wallet. See [#24142 (comment)](https://github.com/bitcoin/bitcoin/pull/24142#issuecomment-1024145620)

That use case is what I call generalized sweep. Sweep can be viewed as spending all UTXOs in a given list without needing to specify the output amount. Whether that list is all the UTXOs in the wallet or a specific list of some UTXOs doesn't matter. So sweep can be extended to cover your use case by allowing for inputs to be specified.

>  but it seems like it would be generally useful whenever you are sending money to one of your own wallets, or exchange accounts, or paying for any service that can be incrementally topped up.

Can you describe why SFFO would be used in those cases? @Xekyo and I have had discussions with many people about SFFO use cases and even though the concept of "send no more than X" usually comes up, no one can express why that behavior would ever actually be useful. Just looking through many of the issues about SFFO show that its primary use case is to sweep an entire wallet, and sometimes spend just a preset list of coins without having to calculate the fee manually. Both of these cases can be covered by a sweep function without needing to maintain SFFO in coin selection logic.

> I know one of achow101's recent PR was adding more complexity to subtract from output implementation to try to do something to help sweeping, but I don't think the original (current?) semantics inherently had to add much complexity. It is true the code has been complex at different points but I think that was mostly a result of code shittiness and duplication, which have generally improved recently.

I find that SFFO makes it much harder to reason about our coin selection because we now use effective values for selection, but with SFFO, we don't actually want the effective values.

luke-jr commented:
Another SFFO use case is when your recipient is legitimately taking on the cost of business. For example, it can be used when sending a loan.

ryanofsky commented:
> > but it seems like it would be generally useful whenever you are sending money to one of your own wallets, or exchange accounts, or paying for any service that can be incrementally topped up.
> 
> Can you describe why SFFO would be used in those cases? @Xekyo and I have had discussions with many people about SFFO use cases and even though the concept of "send no more than X" usually comes up, no one can express why that behavior would ever actually be useful.

I think, mostly, it is useful *conceptually* as a way to simplify creating transactions. It's a way of expressing the *intention* behind a transaction, and letting the implementation take care of the details, instead of being having to think about what the transaction will look like, and choose between sweep and send APIs to create it.

Also the concept here is not exactly "spend no more than X." If there are two amounts associated with transaction: X is amount spent by sender, Y is amount received by receiver, then option lets you choose between fixing X and letting Y vary, or fixing Y and letting X vary. If you have 12 BTC, and want to budget exactly 1 BTC per month to spend on a service, the option let you do that. And in general, it lets you just say what your intention is instead of having to think at a lower level.

Additionally, there are there is my main practical concern:

- This is duplicating existing CreateTransaction(Internal) logic instead of just calling CreateTransaction with the right options. The claim is that this will be a code simplification, but the followup PR #24142 is +588/−148 lines. And this is while REMOVING features which I think we agree are useful: providing GUI support for sweeping, and being able to select coins to sweep. The new code and tests are going to grow and become even more complex after adding these features. 

My medium level practical concern:

- Deprecating subtract from output in #24142 presumably is going to break existing workflows, and there is doesn't seem to be a good release notes or FAQ style item saying what that the problem is with subtract from output, why its removal is justified, and how to transition to the sweep API.

And my mini practical concern:

- This sweep API might be less safe than send APIs because it doesn't force you to specify amount you are trying to send. It's easier to fat-finger by accidentally sweeping the wrong wallet (or when manual coin selection is added) sweeping the wrong coin and spending an amount larger than you intended. Maybe this is not a very big concern. If it is a concern, it could also be addresed by adding a mandatory amount argument requiring you to specify total amount of the sweep or an override value like `"unchecked"` or `"yolo"`

ryanofsky commented:
> I find that SFFO makes it much harder to reason about our coin selection because we now use effective values for selection, but with SFFO, we don't actually want the effective values.

Could you explain this more? There is a [`m_subtract_fee_outputs ? m_value : effective_value`](https://github.com/bitcoin/bitcoin/blob/5b4b8f76f3ae11064d4aa3ac157558e364751fd2/src/wallet/coinselection.cpp#L355-L358) line but that hardly seems terrible

Sjors commented:
This discussion is getting a little mixed between here and https://github.com/bitcoin/bitcoin/pull/24142#issuecomment-1024417067. I don't need _automatic_ coin selection for the use case I describe above. I do still need the feature to work for _manual_ coin selection. Having to use a separate "sweep" RPC for that seems a bit odd, especially when it needs to support most of the `send` kitchen sink, like using hardware wallets, specifying the fee rate, RBF, etc. And it still has to work in the GUI send screen, which touches much of the same CWallet code.

achow101 commented:
> > I find that SFFO makes it much harder to reason about our coin selection because we now use effective values for selection, but with SFFO, we don't actually want the effective values.
> 
> Could you explain this more? There is a [`m_subtract_fee_outputs ? m_value : effective_value`](https://github.com/bitcoin/bitcoin/blob/5b4b8f76f3ae11064d4aa3ac157558e364751fd2/src/wallet/coinselection.cpp#L355-L358) line but that hardly seems terrible
> 
> EDIT: I see there is a counterpart to this in `CreateTransactionInternal` where the option is used to [set](https://github.com/bitcoin/bitcoin/blob/5b4b8f76f3ae11064d4aa3ac157558e364751fd2/src/wallet/spend.cpp#L733-L735) [tx_noinput_size](https://github.com/bitcoin/bitcoin/blob/5b4b8f76f3ae11064d4aa3ac157558e364751fd2/src/wallet/spend.cpp#L741-L743). This seems pretty straightforward as well. Is this the extent of the problem or is there more under the surface?

It makes the reasoning harder because the new coin selection code is almost entirely predicated on using effective values, particularly positive effective values. For BnB and SRD, we assume there are no negative ev inputs. For Knapsack and SRD, we assume we will make change and require a minimum change value. SFFO violates these assumptions, and that leads to bugs like #23026 (which happens to be a sweep use case).

> and choose between sweep and send APIs to create it.
> ...
> This sweep API might be less safe than send APIs because it doesn't force you to specify amount you are trying to send. It's easier to fat-finger by accidentally sweeping the wrong wallet (or when manual coin selection is added) sweeping the wrong coin and spending an amount larger than you intended. Maybe this is not a very big concern. If it is a concern, it could also be addresed by adding a mandatory amount argument requiring you to specify total amount of the sweep or an override value like `"unchecked"` or `"yolo"`

There was some discussion about combining this with send where there would be some option for sweeping and some interpretation of magic values, but we felt that would end up being too clunky to use correctly so it was abandoned.

> This is duplicating existing CreateTransaction(Internal) logic instead of just calling CreateTransaction with the right options. The claim is that this will be a code simplification, but the followup PR [Deprecate SubtractFeeFromOutputs #24142](https://github.com/bitcoin/bitcoin/pull/24142) is +588/−148 lines. 

The followup only has more additions because it is adding `-deprecatedrpc=sffo`. It does not remove the behavior yet.

> And this is while REMOVING features which I think we agree are useful: providing GUI support for sweeping, and being able to select coins to sweep.

There will be GUI support in the future.

> The new code and tests are going to grow and become even more complex after adding these features.

This new code is unlikely to be expanded further given it's limited scope. However I agree there could be some refactors for deduplication between sweep and CreateTransactionInternal.

> Deprecating subtract from output in [Deprecate SubtractFeeFromOutputs #24142](https://github.com/bitcoin/bitcoin/pull/24142) presumably is going to break existing workflows, and there doesn't seem to be a good release notes or FAQ style item saying what that the problem is with subtract from output, why its removal is justified, and how to transition to the sweep API.

The same could be said about every other time `-deprecatedrpc` has been used. Needing docs is not a reason to block a deprecation.

Xekyo commented:
> I think, mostly, it is useful _conceptually_ as a way to simplify creating transactions. It's a way of expressing the _intention_ behind a transaction, and letting the implementation take care of the details, instead of being having to think about what the transaction will look like, and choose between sweep and send APIs to create it.

After talking to some people in the last week and reading all comments on #24142 and here, I think I have a better understanding of how to distinguish the two domains of use cases between sweep and SFFO:

- SFFO: Send to a recipient using a given budget deducting fees
- Sweep: Send sum of inputs to recipient(s) deducting fees

We realize now, that SFFO is used more widely than we anticipated, and that the non-sweeping case is not covered by sweep. The problems and bugs occur when people use SFFO to specify the full wallet balance in order to arrive at a sweep functionality. Therefore, I do think that sweep is still useful.

> * This is duplicating existing CreateTransaction(Internal) logic instead of just calling CreateTransaction with the right options. The claim is that this will be a code simplification, but the followup PR #24142 is +588/−148 lines. And this is while REMOVING features which I think we agree are useful: providing GUI support for sweeping, and being able to select coins to sweep. The new code and tests are going to grow and become even more complex after adding these features.

We looked into this, but found that we need to know the feerate to determine which inputs are eligible, need the inputs to be able to determine the receiver amounts, and when we have all the parameters to call `CreateTransaction(…)`, we already have used all of the duplicated code.

Xekyo commented:
Added the `inputs` option to sweep in 2664d3a.

Sjors commented:
I'm still a bit worried about the amount of duplicated code here, also in the RPC parameter parsing.

Can you add some test coverage for external signer use? The easiest way is probably to tweak `wallet_signer.py` such that the `send` example results in exactly the same coin selection as the new `sweep` example. In that case you can just duplicate the `hww.send` line and compare `res[hex]` between the two calls.

If you end up with two separate transactions because the RPC calls can't produce identical results, then it's useful to know the mechanism that test uses: it has one wallet with private keys and a watch-only clone. A PSBT is generated on the watch-only wallet. The private key enabled wallet is then used to sign it, with the result stored in the file system. Once the test node calls `mocks/signer.py` it reads the stored signed PSBT and returns it, thus faking a hardware wallet that actually signed something. This can of course be done for two separate PSBT's with some refactoring, but having an identical PSBT to sign is easier.

Xekyo commented:
Renamed the proposed RPC `sweepwallet` to clarify that we are operating on the wallet's UTXO pool rather than a wallet-foreign private key as "sweep" may imply.

Rewrote the commit description.

@Sjors: I will look into your feedback to reduce the code duplication and test external signer use.

murchandamus committed:
Extract prevention of outdated option names

This will be reused in `sendall` so we extract it to avoid
duplication.

murchandamus committed:
Extract interpretation of fee estimation arguments

This will be reused in `sendall`, so we extract a method to prevent
duplication.

murchandamus committed:
Extract FinishTransaction from send()

The final step of send either produces a PSBT or the final transaction.
We extract these steps to a new helper function `FinishTransaction()` to
reuse them in `sendall`.

Xekyo commented:
Deduplicated the parameter processing, still working on the test in `wallet_signer.py`.

Xekyo commented:
Now with `wallet_signer.py` test.

Xekyo commented:
Now with Release Notes.

jonatack commented on doc/release-notes-24118.md:
```diff
@@ -0,0 +1,24 @@
+Add `sweepwallet` RPC
```
b2b1d6b

- I think you can add your release note directly to the file `doc/release-notes.md` in this section:

```md
New RPCs
--------
```

- s/without unspecified/without a specified/
- add `(#24118)` to the end

<details><summary>suggested edit with fixups to work from</summary><p>

```text
The `sweepwallet` RPC spends the full balance of the provided UTXOs to one or
more receivers, without creating change.

By default, `sweepwallet` empties the wallet completely, leaving no UTXOs
behind. The optional `send_max` option allows skipping uneconomic UTXOs, thereby
maximizing the received amount.

The `sweepwallet` RPC can process a combination of receiver addresses, with and
without amounts specified. The specified amounts are paid first, and the
remainder is split among the receivers having unspecified amounts. At least one
address must be provided without a specified amount to receive the balance left
after fees.

Instead of the complete UTXO pool, the call can be used to create a transaction
from a specific UTXO set. We recommend using `sweepwallet` to empty wallets
or to spend specific UTXOs in full.

The `sweepwallet` RPC therefore provides a less cumbersome way of spending
specific UTXOs or emptying wallets than subtract fee from output/amount (SFFO).
If the user wishes to specify a budget rather than a set of UTXOs to delimit a
transaction, they should continue to use SFFO. (#24118)
```
</p></details>

jonatack commented on src/wallet/rpc/spend.cpp:
```diff
@@ -7,11 +7,13 @@
 #include <policy/policy.h>
 #include <rpc/rawtransaction_util.h>
 #include <rpc/util.h>
+#include <util/rbf.h>
```
b2b1d6b nit, sort

jonatack commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1159,248 @@ RPCHelpMan send()
+        "\nSpend all (or specific) confirmed UTXOs in the wallet to one or more recipients.\n"
+        "Unconfirmed inbound UTXOs and locked UTXOs will not be spent. Sweepwallet will respect the wallet flag for avoid_reuse.\n",
+        {
+            {"receivers", RPCArg::Type::ARR, RPCArg::Optional::NO, "The destinations of the sweep, each address may only appear once.\n"
```
b2b1d6b use a semi-colon here (";") or a new sentence
```suggestion
            {"receivers", RPCArg::Type::ARR, RPCArg::Optional::NO, "The sweep destinations. Each address may only appear once.\n"
```

Xekyo replied:
Thanks, sounds much better.

jonatack commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1159,248 @@ RPCHelpMan send()
+                        {"locktime", RPCArg::Type::NUM, RPCArg::Default{0}, "Raw locktime. Non-0 value also locktime-activates inputs"},
+                        {"lock_unspents", RPCArg::Type::BOOL, RPCArg::Default{false}, "Lock selected unspent outputs"},
+                        {"psbt", RPCArg::Type::BOOL,  RPCArg::DefaultHint{"automatic"}, "Always return a PSBT, implies add_to_wallet=false."},
+                        {"sendmax", RPCArg::Type::BOOL, RPCArg::Default{false}, "When true, only sweep UTXOs that can pay for their own fees to maximize the output amount. When 'false' (default), no UTXO is left behind. Sendmax is incompatible with providing specific inputs."},
```
b2b1d6b per the developer notes, snake case is used for new RPC arguments: `send_max`?

Xekyo replied:
Done

jonatack commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1159,248 @@ RPCHelpMan send()
+
+            if (lock_unspents) {
+                for (const CTxIn& txin : rawTx.vin) {
+                    pwallet->LockCoin(txin.prevout);
```
b2b1d6b minor suggestions if you retouch and are so inclined

<details><summary>diff</summary><p>

```diff
@@ -1353,17 +1353,11 @@ RPCHelpMan sweepwallet()
             }
 
             CCoinControl coin_control;
-
-            SetFeeEstimateMode(*pwallet, coin_control, options["conf_target"], options["estimate_mode"], options["fee_rate"], /* override_min_fee */ false);
+            SetFeeEstimateMode(*pwallet, coin_control, options["conf_target"], options["estimate_mode"], options["fee_rate"], /*override_min_fee=*/false);
 
             coin_control.fAllowWatchOnly = ParseIncludeWatchonly(options["include_watching"], *pwallet);
 
-            bool lock_unspents = false;
-            if (options.exists("lock_unspents")) {
-                lock_unspents = options["lock_unspents"].get_bool();
-            }
-
-            bool rbf = options.exists("replaceable") ? options["replaceable"].get_bool() : pwallet->m_signal_rbf;
+            const bool rbf{options.exists("replaceable") ? options["replaceable"].get_bool() : pwallet->m_signal_rbf};
 
             FeeCalculation fee_calc_out;
             CFeeRate fee_rate = GetMinimumFeeRate(*pwallet, coin_control, &fee_calc_out);
@@ -1422,7 +1416,7 @@ RPCHelpMan sweepwallet()
                 }
             }
 
-            CAmount output_amounts_claimed(0);
+            CAmount output_amounts_claimed{0};
             for (CTxOut out : rawTx.vout) {
                 output_amounts_claimed += out.nValue;
             }
@@ -1431,12 +1425,12 @@ RPCHelpMan sweepwallet()
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Assigned more value to outputs than available funds.");
             }
 
-            CAmount remainder = effective_value - output_amounts_claimed;
+            const CAmount remainder{effective_value - output_amounts_claimed};
             if (remainder < 0) {
                 throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient funds for fees after creating specified outputs.");
             }
 
-            CAmount per_output_without_amount = remainder / addresses_without_amount.size();
+            const CAmount per_output_without_amount = remainder / addresses_without_amount.size();
 
             bool gave_remaining_to_first = false;
             for (CTxOut& out : rawTx.vout) {
@@ -1461,6 +1455,7 @@ RPCHelpMan sweepwallet()
                 }
             }
 
+            const bool lock_unspents{options.exists("lock_unspents") ? options["lock_unspents"].get_bool() : false};
             if (lock_unspents) {
                 for (const CTxIn& txin : rawTx.vin) {
                     pwallet->LockCoin(txin.prevout);
```
</p></details>

Xekyo replied:
I think I got all these.

jonatack reviewed (commented):
WIP review, looks mostly good so far. A few minor comments if you retouch.

Xekyo commented:
• Addressed @jonatack's comments
• Added test for duplicate destinations
• Added test for sweeping to multiple recipients sharing the remainder
• Added `@cleanup` decorator to tests
• Cleaned up extremely long lines in tests

Open questions: 
• Should this new RPC only allow one method (either options or positional args) to provide `fee_rate`, `feeconf_target` and `estimate_mode`?

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -50,6 +52,91 @@ static void ParseRecipients(const UniValue& address_amounts, const UniValue& sub
+    }
+}
+
+static UniValue FinishTransaction(const std::shared_ptr<CWallet> pwallet, const UniValue& options, const CMutableTransaction& rawTx) {
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1159,248 @@ RPCHelpMan send()
-            if (options.exists("subtractFeeFromOutputs")) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, "Use subtract_fee_from_outputs");
-            }
+            ParseFeeEstimationInstructions(/*conf_target*/ request.params[1], /*estimate_mode*/ request.params[2], /*fee_rate*/ request.params[3], options);
```
Got em.

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -50,6 +52,91 @@ static void ParseRecipients(const UniValue& address_amounts, const UniValue& sub
+    return result;
+}
+
+static void PreventOutdatedOptions(const UniValue& options) {
```
Done.

luke-jr commented:
Hate to nitpick the name, but "sweepwallet" to me sounds too easily confused with sweeping funds *into* the wallet.

jonatack reviewed (commented):
Looks like `test/lint/lint-python.sh` needs appeasing
```
test/functional/wallet_sweep.py:28:1: E115 expected an indented block (comment)
```

jonatack commented on doc/release-notes.md:
```diff
@@ -126,6 +126,30 @@ Updated RPCs
+  The `sweepwallet` RPC therefore provides a less cumbersome way of spending
+  specific UTXOs or emptying wallets than subtract fee from output/amount (SFFO).
+  If the user wishes to specify a budget rather than a set of UTXOs to delimit a
+  transaction, they should continue to use SFFO. (#24118)
```
I wonder if the release note isn't a bit long and whether the how-to info would be best documented elsewhere more permanently like in the sweepwallet help or in one of the doc/ files.

Xekyo replied:
I rewrote the release notes.

Xekyo commented:
> Hate to nitpick the name, but "sweepwallet" to me sounds too easily confused with sweeping funds _into_ the wallet.

I changed it from `sweep` after multiple people stated that they first thought it was a feature to sweep a private key into the wallet. I'm open to suggestions, if you have an idea for a better name.

luke-jr commented on doc/release-notes.md:
```diff
@@ -126,6 +126,30 @@ Updated RPCs
 
+- Add `sweepwallet` RPC
+
+  The `sweepwallet` RPC spends some given UTXOs' complete balance to one or more
```
nit: s/balance/value

luke-jr commented:
idk, I guess if you don't mind ignoring the advanced use cases in naming, "emptywallet" might work.

flack commented:
or maybe "drainwallet". "emptywallet" sounds a lot like empty wallet, i.e. it's not immediately clear that "empty" is a verb in this case

Xekyo commented on test/functional/test_runner.py:
```diff
@@ -278,6 +278,8 @@
     'wallet_send.py --legacy-wallet',
     'wallet_send.py --descriptors',
+    'wallet_sweep.py --legacy-wallet',
+    'wallet_sweep.py --descriptors',
```
I see that some other wallet related calls do start with `rpc_…` as well, such as `rpc_rawtransaction.py`, `rpc_fundrawtransaction.py`, `rpc_signrawtransaction.py`, and `rpc_psbt.py`, but all the other send related calls are under `wallet_…`. I'm kinda thinking that it fits the pattern as it is currently. It should probably be `wallet_sweepwallet.py`, though.

Xekyo replied:
Renamed to `wallet_sweepwallet.py`.

Xekyo commented:
Moved the release notes back to their own file to avoid merge conflict (I think I had misunderstood some feedback I got about release notes).

Xekyo commented:
Renamed RPC to `sendall`, after discussion with other contributors.

achow101 commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1168,6 +1181,229 @@ RPCHelpMan send()
+            }
+
+            CMutableTransaction rawTx = ConstructTransaction(options["inputs"], receiver_key_value_pairs, options["locktime"], rbf);
+            LOCK(pwallet->cs_wallet); // Lock automatically released at end of function
```
in 5a663b2a676ed7ec24d5d4b1690d2639aeebab86 "Add sendall RPC née sweep"

nit: The comment on this line is unnecessary.

Xekyo replied:
Thanks, removed

achow101 commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1168,6 +1181,229 @@ RPCHelpMan send()
+                    total_input_value += tx->tx->vout[input.prevout.n].nValue;
+                }
+            } else {
+                AvailableCoins(*pwallet, all_the_utxos, &coin_control, /*spend 0-value utxos=*/0);
```
in 5a663b2a676ed7ec24d5d4b1690d2639aeebab86 "Add sendall RPC née sweep"

nit: the inline comment should have the parameter name

```suggestion
                AvailableCoins(*pwallet, all_the_utxos, &coin_control, /*nMinimumAmount=*/0);
```

Xekyo replied:
Thanks, updated

achow101 commented on test/functional/wallet_sendall.py:
```diff
@@ -0,0 +1,291 @@
+        self.generate(self.nodes[0], 1)
+        assert_greater_than(self.wallet.getbalances()["mine"]["trusted"], 0)
+
+        # Clean up remaining UTXO
+        self.wallet.sendall(receivers=[self.remainder_target])
+        self.generate(self.nodes[0], 1)
+        assert_equal(0, self.wallet.getbalances()["mine"]["trusted"]) # wallet is empty
```
In 5a663b2a676ed7ec24d5d4b1690d2639aeebab86 "Add sendall RPC née sweep"

This cleanup is no longer necessary with the `@cleanup` decorator.

Xekyo replied:
Indeed. I removed the duplicate cleanup. Thanks.

Sjors commented on test/functional/wallet_sendall.py:
```diff
@@ -0,0 +1,286 @@
+        foreign_utxo = self.def_wallet.listunspent()[0]
+        assert_raises_rpc_error(-8, "Input not found. UTXO ({}:{}) is not part of wallet.".format(foreign_utxo["txid"],
+            foreign_utxo["vout"]), self.wallet.sendall, receivers=[self.remainder_target],
+            options={"inputs": [foreign_utxo]})
```
Some of the times when I run the test it raises "Total value of UTXO pool too low to pay for transaction." and then during the handling it said "No exception raised" for this line.

Xekyo replied:
I have been able to reproduce this, but I haven't figured out what the issue is, yet.

achow101 replied:
The issue is that `foreign_utxo` might be the change output for a transaction that funded the test wallet. Because the transaction has something belonging to the test wallet, the test wallet knows the transaction and does not detect it to be foreign. I will add a suggestion at the relevant line to fix this issue.

Xekyo replied:
Thanks, fixed working from achow101's suggestion

Sjors reviewed (commented):
Review hint for the first three commits: `git show --color-moved=dimmed-zebra --color-moved-ws=allow-indentation-change`

It would be nice if for a single recipient you can omit the array, like so: `sendall bc1q09...hufdl` (instead of `sendall '["bc1q09...hufdl"]'`). 

The first 3 commits are useful refactors regardless, just in case this PR doesn't make it (ACK on those, but still need to review the meat and potatoes commit).

In the description of 9bda5234bc41b7477b0224930177c6dfcd8cb9fc maybe replace "given" with "specific".

Let's add a quick warning to the help text:

> If your wallet contains many small inputs, either because it received tiny payments or as a result of accumulating change, consider using "send_max" to exclude those inputs that are worth less than the fees needed to spend them.

---

Thoughts that can wait for followups...

Name-wise `sendall` implies that you can't send specific UTXO's. We can think of a better name later, since it's marked experimental. One suggestion: `sendcoins`. That said, I doubt anyone uses the RPC to _manually_ perform coin selection. The GUI or some external tool is far easier to use for that. So that leaves the use case of emptying out the wallet, and for that the current name is very accurate.

We could potentially make sending specific coins less tedious on the RPC, but that's outside the scope of this PR. Ideally you would want a new RPC call that returns a concise list of outpoints with their amount, like so:

```
16a9...97f:7 0.000100
34fe...e8d:0 0.00200
```

We could even use transaction short ids, if there's no duplicate in any given wallet (return the full id if there is).

And then rather than the tedious options dictionary with an object for each input, we'd add one argument which takes a simple array of strings, so you can do `sendall bc1q09...hufdl null unset 1.1 '["16a9...97f:7"]'` in order to spend the first coin. Still not super pretty, but when copy-pasted and adjusted from the help example, it should be easy enough to use.

To burn dust we could add an option `burn` that, if zero addresses are provided, will send to `OP_RETURN`. That way you might sweep economic coins with a slightly higher fee and then make a burn transaction for the rest at 1 sat/vbyte (or 0 if you have a way to get those mined).

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1168,6 +1181,229 @@ RPCHelpMan send()
+                for (const CTxIn& txin : rawTx.vin) {
+                    pwallet->LockCoin(txin.prevout);
+                }
+            }
```
General organization question: Did you consider putting some of this in a `CWallet` member function? A lot of it, like grabbing available coins and checking the effective values, locking unspents, etc. seem to be things that the wallet should take care of internally, rather than in the rpc handler code.

Sjors replied:
That seems like a good idea, but maybe too big a refactor for this PR?

glozow replied:
This is all added code?

Sjors replied:
Oops yes, this is not part of the refactoring commits.

Xekyo replied:
That sounds like a good idea. Especially getting subsets of a wallet's UTXO pool does seem like a good candidate. I'll consult with my wallet expert on this matter.

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1168,6 +1181,229 @@ RPCHelpMan send()
+                    total_input_value += tx->tx->vout[input.prevout.n].nValue;
+                }
+            } else {
+                AvailableCoins(*pwallet, all_the_utxos, &coin_control, /*nMinimumAmount=*/0);
```
Given that `AvailableCoins` filters much more than based on effective value, I'm wondering what the expected behavior *should* be for these types of UTXOs:

- UTXOs that are not yet final or are immature coinbases 
- UTXOs for which the address has been marked as already used
- outputs from replacement transactions
etc.

Maybe it's fine that these are excluded, but just want to clarify that this behavior is what we want. Also leads me to wonder if the RPC should be returning a list of UTXOs that weren't included in the constructed transaction or something, but maybe that's overkill.

Xekyo replied:
This RPC will only operate on confirmed spendable UTXOs. It also respects the wallet-wide `avoid_reuse` configuration.

> * UTXOs that are not yet final or are immature coinbases

I would expect them to get ignored since they're not spendable at this time.

> * UTXOs for which the address has been marked as already used

If address reuse has been disabled, it should ignore funds on previously used addresses, if it only permits spending all of the funds on one address at once, it should either use all or none, and if it's not restricted, it should just spend either all or the economically viable UTXOs as determined by the options on this RPC.

> * outputs from replacement transactions

We don't spend unconfirmed UTXOs in this call.

glozow reviewed (commented):
Still reviewing, but posting some questions I had about the expected behavior

ishaanam committed:
add tests for no recipient and using send_max while inputs are specified

stickies-v commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+                    {
+                        {"add_to_wallet", RPCArg::Type::BOOL, RPCArg::Default{true}, "When false, returns a serialized transaction which will not be added to the wallet or broadcast"},
+                        {"fee_rate", RPCArg::Type::AMOUNT, RPCArg::DefaultHint{"not set, fall back to wallet fee estimation"}, "Specify a fee rate in " + CURRENCY_ATOM + "/vB."},
+                        {"include_watching", RPCArg::Type::BOOL, RPCArg::DefaultHint{"true for watch-only wallets, otherwise false"}, "Also select inputs which are watch only.\n"
```
Consistency nit
```suggestion
                        {"include_watching", RPCArg::Type::BOOL, RPCArg::DefaultHint{"true for watch-only wallets, otherwise false"}, "Also select inputs which are watch-only.\n"
```

Xekyo replied:
Thanks

stickies-v commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+                "options", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED_NAMED_ARG, "",
+                Cat<std::vector<RPCArg>>(
+                    {
+                        {"add_to_wallet", RPCArg::Type::BOOL, RPCArg::Default{true}, "When false, returns a serialized transaction which will not be added to the wallet or broadcast"},
```
Wording nit/suggestion:
```suggestion
                        {"add_to_wallet", RPCArg::Type::BOOL, RPCArg::Default{true}, "When false, returns the serialized transaction without broadcasting or adding it to the wallet"},
```

Xekyo replied:
That's better, thanks

stickies-v reviewed (commented):
Approach ACK 9bda523

A couple of nits and wording suggestions, all of them optional. I support the `sendall` name, I think it's neutral and well describes what's happening.

Xekyo commented:
@Sjors: Thanks for the thoughtful suggestions. I've adopted your suggestions for phrasing and discussed whether overloading the receivers/recipients array with a single string with some other developers. While I agree that it would ease the (probably) most likely use case of sweeping everything into a single address, some people didn't like the idea of allowing multiple types for one parameter.

@glozow, @stickies-v: Thank you for the review, I'm working on answers to each of your comments.

Xekyo commented on doc/release-notes-24118.md:
```diff
@@ -0,0 +1,10 @@
+  everything in the wallet leaving no UTXOs behind. `sendall` is
+  useful to empty wallets or to create a changeless payment from select
+  UTXOs. Continue to use the "subtract fee from output" option to create
+  transactions delimited via a budget (in contrast to a concrete set of
```
Thanks for the suggestions. I'm changing the sentence to:

> When creating a payment from a specific amount for which the recipient incurs the transaction fee, continue to use the   `subtractfeefromamount` option via the `send`, `sendtoaddress`, or `sendmany` RPCs.

Xekyo commented on src/rpc/client.cpp:
```diff
@@ -142,6 +142,10 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "send", 1, "conf_target" },
     { "send", 3, "fee_rate"},
     { "send", 4, "options" },
+    { "sendall", 0, "receivers" },
```
Changing receiver to recipient everywhere.

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -51,6 +51,28 @@ static void ParseRecipients(const UniValue& address_amounts, const UniValue& sub
     }
 }
 
+static void ParseFeeEstimationInstructions(const UniValue& positional_conf_target, const UniValue& positional_estimate_mode, const UniValue& positional_fee_rate, UniValue& options)
```
Dropping the `positional_` and renaming function to `InterpretFeeEstimationInstructions`

Xekyo commented:
Sorry if this is confusing, but I've started responding to comments, but haven't finished making all the code changes and rebasing yet (gotta switch devices and didn't want to lose what I've already written).

Xekyo commented on doc/release-notes-24118.md:
```diff
@@ -0,0 +1,10 @@
+
+- The `sendall` RPC spends given UTXOs to one or more receivers
+  without creating change. By default, the `sendall` RPC will send
+  everything in the wallet leaving no UTXOs behind. `sendall` is
```
Adopted your suggestion.

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+
+            std::set<std::string> addresses_without_amount;
+            UniValue receiver_key_value_pairs(UniValue::VARR);
+            const UniValue& receivers = request.params[0];
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -50,6 +53,93 @@ static void ParseRecipients(const UniValue& address_amounts, const UniValue& sub
+    }
+
+    if (complete) {
+        std::string hex = EncodeHexTx(CTransaction(mtx));
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+                }, true
+            );
+
+            std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+            const bool rbf{options.exists("replaceable") ? options["replaceable"].get_bool() : pwallet->m_signal_rbf};
+
+            FeeCalculation fee_calc_out;
+            CFeeRate fee_rate = GetMinimumFeeRate(*pwallet, coin_control, &fee_calc_out);
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
 
-            CMutableTransaction mtx;
-            complete = FinalizeAndExtractPSBT(psbtx, mtx);
+            CMutableTransaction rawTx = ConstructTransaction(options["inputs"], receiver_key_value_pairs, options["locktime"], rbf);
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+                    if (pwallet->IsSpent(input.prevout.hash, input.prevout.n)) {
+                        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Input not available. UTXO (%s:%d) was already spent.", input.prevout.hash.ToString(), input.prevout.n));
+                    }
+                    const CWalletTx* tx = pwallet->GetWalletTx(input.prevout.hash);
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+            // estimate final size of tx
+            TxSize tx_size = CalculateMaximumSignedTxSize(CTransaction(rawTx), pwallet.get());
+            CAmount fee_from_size = fee_rate.GetFee(tx_size.vsize);
+            CAmount effective_value = total_input_value - fee_from_size;
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+                throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient funds for fees after creating specified outputs.");
+            }
+
+            const CAmount per_output_without_amount = remainder / addresses_without_amount.size();
```
Done

Xekyo replied:
This one actually doesn't like being a brace initialization:

```
wallet/rpc/spend.cpp: In lambda function:
wallet/rpc/spend.cpp:1371:63: warning: narrowing conversion of ‘(((long unsigned int)((long int)remainder)) / addresses_without_amount.std::set<std::__cxx11::basic_string<char> >::size())’ from ‘long unsigned int’ to ‘CAmount’ {aka ‘long int’} inside { } [-Wnarrowing]
             const CAmount per_output_without_amount{remainder / addresses_without_amount.size()};
                                                     ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Xekyo replied:
Fixed by casting the size to `(long)` explicitly.

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+            for (CTxOut& out : rawTx.vout) {
+                CTxDestination dest;
+                ExtractDestination(out.scriptPubKey, dest);
+                std::string addr = EncodeDestination(dest);
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+    return RPCHelpMan{"sendall",
+        "EXPERIMENTAL warning: this call may be changed in future releases.\n"
+        "\nSpend the value of all (or specific) confirmed UTXOs in the wallet to one or more recipients.\n"
+        "Unconfirmed inbound UTXOs and locked UTXOs will not be spent. Sendall will respect the wallet flag for avoid_reuse.\n",
```
Amended

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+            UniValue receiver_key_value_pairs(UniValue::VARR);
+            const UniValue& receivers = request.params[0];
+            for (unsigned int i = 0; i < receivers.size(); ++i) {
+                const UniValue& receiver = receivers[i];
```
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1060,101 +1162,244 @@ RPCHelpMan send()
+
+            const CAmount per_output_without_amount = remainder / addresses_without_amount.size();
+
+            bool gave_remaining_to_first = false;
```
Done

Xekyo commented:
Addressed most comments by Sjors, Gloria and stickies-v.

Still looking into/open for feedback on:
- @sjors's suggestion of overloading recipients parameter to take a single string in the case of a single recipient instead of an array
- The inconsistent bug @Sjors noticed in the  `sendall_fails_on_missing_input` test
- @glozow's suggestion of moving parts of the UTXO filtering into CWallet
- @stickies-v's suggestion to align the error messages

achow101 commented on test/functional/wallet_signer.py:
```diff
@@ -192,6 +192,12 @@ def test_valid_signer(self):
 
+        self.log.info('Test sendall using hww1')
+
+        res = hww.sendall(receivers=[{dest:0.5}, hww.getrawchangeaddress()],options={"add_to_wallet": False})
```
In 5c0a1b330294d40112af79af19414b913d0dce8c "Add sendall RPC née sweep"

```suggestion
        res = hww.sendall(recipients=[{dest:0.5}, hww.getrawchangeaddress()],options={"add_to_wallet": False})
```

Xekyo replied:
Fixed, thank you

achow101 commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1168,6 +1181,230 @@ RPCHelpMan send()
+                        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Input not available. UTXO (%s:%d) was already spent.", input.prevout.hash.ToString(), input.prevout.n));
+                    }
+                    const CWalletTx* tx{pwallet->GetWalletTx(input.prevout.hash)};
+                    if (!tx) {
```
We may accidentally allow an external UTXO here if the transaction had both outputs that belonged to us, and outputs that did not. So we should check that the specified output itself actually belongs to this wallet.

```suggestion
                    if (!tx || pwallet->IsMine(tx->tx->vout[input.prevout.n]) != ISMINE_SPENDABLE | (coin_control.fAllowWatchOnly ? ISMINE_WATCHONLY : 0)) {
```

(Did not test it, but something similar to that should work. The output needs to be in the wallet, and we should also respect `coin_control.fAllowWatchonly`)

Fixes https://github.com/bitcoin/bitcoin/pull/24118/commits/5c0a1b330294d40112af79af19414b913d0dce8c#r809858573

Xekyo replied:
Thanks! I've used 
`if (!tx || pwallet->IsMine(tx->tx->vout[input.prevout.n]) != (coin_control.fAllowWatchOnly ? ISMINE_ALL : ISMINE_SPENDABLE))`

murchandamus committed:
Elaborate error messages for outdated options

Xekyo commented on test/functional/wallet_sendall.py:
```diff
@@ -0,0 +1,286 @@
+        for eo in expected_outputs:
+            self.assert_tx_has_output(tx, eo["address"], eo["value"])
+
+    def add_uxtos(self, amounts):
```
Fixed everywhere, thanks

Xekyo replied:
:scream: Fixed

Xekyo commented:
Added a second test provided by @ishaanam. Thanks!

mzumsande commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1235,6 +1248,230 @@ RPCHelpMan send()
+        },
+        RPCExamples{""
+        "\nSpend all UTXOs from the wallet with a fee rate of 1 " + CURRENCY_ATOM + "/vB using named arguments\n"
+        + HelpExampleCli("-named sendall", "recipients='{\"" + EXAMPLE_ADDRESS[0] + "\"}' fee_rate=1\n") +
```
In the first help example, use [...] instead of {..} for a valid json array.

achow101 commented:
ACK 08be21737e1b38a8171ca216c18a7bbf94511c49

w0xlt reviewed (approved):
re-ACK 08be217

MarcoFalke commented:
Needs rebase?

achow101 commented:
> Needs rebase?

Yes, there's a silent merge conflict.

Xekyo commented:
Rebased to master due to silent merge conflict.

achow101 commented:
re-ACK 78aab643af1c1afbaf82e1269b55a617f7fda6f6

Changes since last were just the rebase to resolving the silent merge conflict with renaming variables.

w0xlt reviewed (approved):
re-ACK 78aab64

achow101 commented:
@sjors @stickies-v @glozow re-review this?

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1179,49 +1243,231 @@ RPCHelpMan send()
 
-            UniValue result(UniValue::VOBJ);
+            // estimate final size of tx
+            TxSize tx_size{CalculateMaximumSignedTxSize(CTransaction(rawTx), pwallet.get())};
+            CAmount fee_from_size{fee_rate.GetFee(tx_size.vsize)};
+            CAmount effective_value{total_input_value - fee_from_size};
```
nit in 3813ba1b196f8dd89de1f8b959c179dd28c05d4e

```suggestion
            const TxSize tx_size{CalculateMaximumSignedTxSize(CTransaction(rawTx), pwallet.get())};
            const CAmount fee_from_size{fee_rate.GetFee(tx_size.vsize)};
            const CAmount effective_value{total_input_value - fee_from_size};
```

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1179,49 +1243,231 @@ RPCHelpMan send()
+            std::vector<COutput> all_the_utxos;
+
+            CAmount total_input_value(0);
+            bool send_max{options.exists("send_max") && options["send_max"].get_bool()};
```
nit in 3813ba1b196f8dd89de1f8b959c179dd28c05d4e
```suggestion
            const bool send_max{options.exists("send_max") && options["send_max"].get_bool()};
```

MarcoFalke replied:
Generally I'd prefer to keep code logic and constants (defaults) separate. Otherwise changing the constant (default) requires changing the code logic. So this should be `bool a{option.exists() ? option.bool() : option_default};`

Xekyo replied:
Changed this line to match @MarcoFalke's suggestion.

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1179,49 +1243,231 @@ RPCHelpMan send()
+                output_amounts_claimed += out.nValue;
+            }
+
+            if (output_amounts_claimed > total_input_value) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Assigned more value to outputs than available funds.");
+            }
+
+            const CAmount remainder{effective_value - output_amounts_claimed};
+            if (remainder < 0) {
+                throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient funds for fees after creating specified outputs.");
+            }
```
What makes one of these `RPC_INVALID_PARAMETER` and the other `RPC_WALLET_INSUFFICIENT_FUNDS` ?

Xekyo replied:
Changed to `RPC_WALLET_INSUFFICIENT_FUNDS`

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1179,49 +1243,231 @@ RPCHelpMan send()
+
+            std::shared_ptr<CWallet> const pwallet{GetWalletForJSONRPCRequest(request)};
+            if (!pwallet) return NullUniValue;
+
```
Do you need to `pwallet->BlockUntilSyncedToCurrentChain();` here?

Xekyo replied:
Yes! Thanks.

glozow reviewed (commented):
code review ACK 78aab643af1c1afbaf82e1269b55a617f7fda6f6, a few non-blocking comments/questions

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -449,31 +450,43 @@ RPCHelpMan settxfee()
 
 // Only includes key documentation where the key is snake_case in all RPC methods. MixedCase keys can be added later.
-static std::vector<RPCArg> FundTxDoc()
+static std::vector<RPCArg> FundTxDoc(bool solving_data = true)
```
Using this default value in this instance replicates the previous behavior.

Xekyo commented:
Addressed all of the comments by @glozow and @MarcoFalke

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -21,7 +21,8 @@
 
 namespace wallet {
-static void ParseRecipients(const UniValue& address_amounts, const UniValue& subtract_fee_outputs, std::vector<CRecipient> &recipients) {
+static void ParseRecipients(const UniValue& address_amounts, const UniValue& subtract_fee_outputs, std::vector<CRecipient> &recipients)
```
Fixed

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -74,6 +74,50 @@ static void InterpretFeeEstimationInstructions(const UniValue& conf_target, cons
+        CTransactionRef tx(MakeTransactionRef(std::move(mtx)));
+        result.pushKV("txid", tx->GetHash().GetHex());
+        if (add_to_wallet && !psbt_opt_in) {
+            pwallet->CommitTransaction(tx, {}, {} /* orderForm */);
```
Fixed

achow101 commented:
re-ACK bb84b7145b31dbfdcb4cf0b9b6e612a57e573993

MarcoFalke commented:
This had 3 ACKs and I checked the range-diff. Did not review myself.

Sjors commented:
Sorry I didn't get around to reviewing this. I'll see if I can test it later.

This remains on my wish list (for a future PR):

> a new RPC call that returns a concise list of outpoints with their amount, like so:
> 
> ```
> 16a9...97f:7 0.000100
> 34fe...e8d:0 0.00200
> ```
> 
> We could even use transaction short ids, if there's no duplicate in any given wallet (return the full id if there is).
